torch/testing/_internal/common_optimizers.py

```diff
@@ -0,0 +1,626 @@
+import functools
+import itertools
+from enum import Enum
+from typing import Any, Dict, List, Tuple, Union
+
+import torch
+from torch import Tensor
+from torch.nn import Parameter
+from torch.optim import (
+    Adadelta,
+    Adagrad,
+    Adam,
+    Adamax,
+    AdamW,
+    ASGD,
+    LBFGS,
+    NAdam,
+    Optimizer,
+    RAdam,
+    RMSprop,
+    Rprop,
+    SGD,
+    SparseAdam,
+)
+from torch.testing._internal.common_methods_invocations import DecorateInfo
+from torch.testing._internal.common_utils import (
+    _TestParametrizer,
+    set_single_threaded_if_parallel_tbb,
+)
+
+
+class OptimizerInput:
+    """Contains args / kwargs to be passed to an optimizer constructor."""
+
+    __slots__ = ["params", "kwargs", "desc"]
+
+    def __init__(
+        self,
+        params: Union[List[Parameter], List[Tensor], Dict[Any, Any]],
+        kwargs: Dict[str, Any],
+        desc: str = "",
+    ):
+        self.params = (
+            params  # Here, params can be a list of Tensors OR param_groups as well.
+        )
+        self.kwargs = kwargs
+        self.desc = desc
+
+
+class OptimizerErrorEnum(Enum):
+    """Enumerates when an error is raised when testing optimizers."""
+
+    CONSTRUCTION_ERROR = 0
+    STEP_ERROR = 1
+
+
+class ErrorOptimizerInput:
+    """
+    An OptimizerInput that will cause the optimizer to throw an error when constructed.
+    Includes the type and string of the resulting error.
+    """
+
+    __slots__ = ["optimizer_error_input", "error_on", "error_type", "error_regex"]
+
+    def __init__(
+        self,
+        optimizer_error_input,
+        *,
+        error_on=OptimizerErrorEnum.CONSTRUCTION_ERROR,
+        error_type=RuntimeError,
+        error_regex="",
+    ):
+        self.optimizer_error_input = optimizer_error_input
+        self.error_on = error_on
+        self.error_type = error_type
+        self.error_regex = error_regex
+
+
+class OptimizerInfo:
+    """Optimizer information to be used in testing."""
+
+    def __init__(
+        self,
+        optim_cls: Optimizer,  # Class object for the Optimizer under test
+        *,
+        # Function to generate optimizer inputs
+        optim_inputs_func,
+        # Implementation specific kwargs the optimizer supports, e.g., fused, foreach, differentiable
+        # We consider capturable to be a base constructor flag since it is implemented across the board.
+        supported_impls: Tuple[str] = ("foreach", "differentiable"),
```

Comment:
It would be great for the comment above to mention every valid value for this flag.

Reply:
wait it does do that

Reply:
"kwargs the optimizer supports, e.g., fused, foreach, differentiable"
Is not the same as
"kwargs the optimizer supports, it must be a subset of fused, foreach, differentiable"